services/sharedchannel/getpostssince.go

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package sharedchannel
+
+import (
+	"github.com/mattermost/mattermost-server/v5/mlog"
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+type sinceResult struct {
+	posts     []*model.Post
+	hasMore   bool
+	nextSince int64
+}
+
+// getPostsSince fetches posts that need to be synchronized with a remote cluster.
+// There is a soft cap on the number of posts that will be synchronized in a single pass (MaxPostsPerSync).
+//
+// There is a special case where multiple posts have the same UpdateAt value. It is vital that this method
+// include all posts within that millisecond so that subsequent calls can use an incremented `since`. If this
+// method were to be called repeatedly with the same `since` value the same records would be returned each time
+// and the sync would never move forward.
+//
+// A boolean is also returned to indicate if there are more posts to be synchronized (true) or not (false).
+func (scs *Service) getPostsSince(channelId string, rc *model.RemoteCluster, since int64) (sinceResult, error) {
+	opts := model.GetPostsSinceForSyncOptions{
+		ChannelId:      channelId,
+		Since:          since,
+		IncludeDeleted: true,
+		Limit:          MaxPostsPerSync + 1, // ask for 1 more than needed to peek at first post in next batch
+	}
+	posts, err := scs.server.GetStore().Post().GetPostsSinceForSync(opts, true)
```

Comment:
Since we are ordering here by just the updateAt in SQL that makes the result non-deterministic since the dbplanner cannot return the same resultsets for records that have the same `UpdateAt`.  That can lead to potentially missed or duplicate syncs, right?

**Records: [A: updateAt = X, B: updateAt = X, C: updateAt = X, D: updateAt = X]**

Limit = 2

If i understand the algo correctly it will go like this:

- GetPostsSinceForSync() return [A,  B] + [C]
- Offset is calculated to 2
- GetPostsSinceForSync() return [D, C, A, B] limit is 1000 with offset 2 essentially the return list is [A, B] so we append to the initial list A and B.

Reply:
Whoa.  I would have expected natural order (order they were inserted) after ORDER BY is exhausted.  That's how Oracle works.  Thanks for the heads up.  I'll add `Id` as a second OrderBy field.

Reply:
Fixed via https://github.com/mattermost/mattermost-server/pull/16962/commits/3824abe805b8718cd6a130f4b15027e368148d69

Reply:
Bello 👌